### PR TITLE
Feature/add name search and refactor

### PR DESCRIPTION
**Pull request by hreis**
### **Summary**
This PR introduces a refactored Card component with enhanced navigation capabilities and adds a comprehensive suite of tests to ensure robustness and reliability.

### **Implementation Details:**
**Search Feature:** Developed an advanced search functionality to allow users to find specific items within the application more efficiently. This includes filtering options and predictive text inputs.

**Search Algorithm:** Implemented a lightweight and optimized search algorithm that indexes items in real-time for faster retrieval, minimizing the application's load time.

**Refactoring of Card:** Improved the Card component's navigation logic to enable more intuitive user interactions and to prepare the component for future feature additions.

State Handling: Updated the way Card handles internal state to improve performance and to align with React's best practices.

**Folder Structure:**
Components: Organized React components in a hierarchical structure that mirrors their usage within the application, separating UI components & __TESTS__ also.

**Test Suite Enhancements:** Extended the testing suite to cover new logic paths and ensure that all user interaction scenarios are adequately tested, following the increased complexity in navigation logic.

**New Feature Implementation Approach**
Implemented the new navigation feature by leveraging the useNavigate hook to programmatically control routing.
Adopted a defensive coding approach to handle potential edge cases, such as missing column data or an undefined navigation URL.
Followed a TDD (Test-Driven Development) process to ensure that all new code was tested before integration.
Documentation
Updated the README to guide new users through the changes and to describe the component's behavior.
Inline comments added to complex logic sections for better maintainability.

**Testing**
Added new test cases to `Card.test.js` to validate both the navigation functionality and the fallback scenarios where navigation is disabled.
Ensured that tests are readable, maintainable, and align with the project's existing testing patterns.
Looking forward to the review and any feedback that could help improve this PR further.

### CODE REVIEW

File: src/__tests__/pages/testUserOverview.tsx [hunk 1]
```diff
@@ -18,8 +18,8 @@ describe('UserOverview', () => {
     it('should render UserOverview', () => {
         render(<UserOverview />);
 
-        expect(screen.getByText('Test User')).toBeInTheDocument();
-        expect(screen.getByText('userName')).toBeInTheDocument();
-        expect(screen.getByText('location')).toBeInTheDocument();
+        const userElements = screen.getAllByText(/Test User/);
+        expect(userElements).toHaveLength(2);
+        expect(screen.getByText(/location/)).toBeInTheDocument();
```
Comment (hreis) [inline src/__tests__/pages/testUserOverview.tsx:23]: The test was updated because the component now may render multiple instances of user information, hence we check for the presence and count of 'Test User' rather than its singular occurrence.

File: src/components/Card/index.tsx [hunk 2]
```diff
@@ -14,33 +14,78 @@ interface Props {
     navigationProps?: UserData | Teams;
 }
 
-const Card = ({
+const Card: React.FC<Props> = ({
     id,
     columns,
-    url,
+    url = '#',
```
Comment (hreis) [inline src/components/Card/index.tsx:20]: ensured that the url has a fallback value

File: src/components/Card/index.tsx [hunk 3]
```diff
@@ -14,33 +14,78 @@ interface Props {
...
+}) => {
     const navigate = useNavigate();
+    const getInitials = (name = '') => {
+        return name
+            .split(' ')
+            .slice(0, 2)
+            .map((word) => word[0])
+            .join('')
+            .toUpperCase();
+    };
```
Comment (hreis) [inline src/components/Card/index.tsx:32]: have the initials in case avatar https don't load correctly. so the layout don't break.

File: src/components/Card/index.tsx [hunk 4]
```diff
@@ -14,33 +14,78 @@ interface Props {
...
+
+    React.useEffect(() => {
+        const img = new Image();
+        img.onload = () => setAvatarUrl(avatarEntry.value);
+        img.onerror = () => setAvatarUrl('');
+        if (avatarEntry && avatarEntry.value) {
+            img.src = avatarEntry.value;
+        }
+    }, [avatarEntry]);
```
Comment (hreis) [inline src/components/Card/index.tsx:59]: Introduced a mechanism to handle avatar loading within the Card component. If the avatar image fails to load, the component will fall back to displaying initials. This improves UI resilience and user experience.

File: src/components/Card/index.tsx [hunk 5]
```diff
@@ -14,33 +14,78 @@ interface Props {
...
             data-testid={`cardContainer-${id}`}
             hasNavigation={hasNavigation}
-            onClick={(e: Event) => {
-                if (hasNavigation) {
-                    navigate(url, {
-                        state: navigationProps,
-                    });
-                }
-                e.preventDefault();
-            }}
```
Comment (hreis) [inline src/components/Card/index.tsx:37]: not good to have logic inside html. So changed for a modular function

File: src/components/Card/index.tsx [hunk 6]
```diff
@@ -14,33 +14,78 @@ interface Props {
...
+                    justifyContent: 'center',
+                    fontSize: '20px',
+                    borderRadius: '50%',
+                }}>
+                    {initials}
+                </div>
+            )}
+            {teamLeadEntry && <p>Team Lead</p>}
+            {displayNameEntry && <p>{displayNameEntry.value}</p>}
+            {locationEntry && <p>{locationEntry.key}: {locationEntry.value}</p>}
```
Comment (hreis) [inline src/components/Card/index.tsx:88]: new style

File: src/components/FilterInput/index.tsx [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+    <StyledFilterInput
+        type="text"
+        placeholder={placeholder}
+        value={value}
+        onChange={onChange}
+    />
+  );
+};
+
+export default FilterInput;
```
Comment (hreis) [inline src/components/FilterInput/index.tsx:22]: create external filter input component so it can be reused wherever the application needs in the future

File: src/components/Header/index.tsx [hunk 8]
```diff
@@ -9,16 +9,27 @@ interface Props {
...
+            // scenario where there's no history
+            navigate('/home');
+        }
+    };
+
     return (
-        <HeaderContainer>
-            <NavigationHeader>
+        <HeaderContainer as="header">
+            <NavigationHeader as="nav">
```
Comment (hreis) [inline src/components/Header/index.tsx:24]: Improved semantic HTML by specifying as="header" and as="nav" for better accessibility and clearer structure

File: src/components/Header/index.tsx [hunk 9]
```diff
@@ -9,16 +9,27 @@ interface Props {
...
     const navigate = useNavigate();
+
+    const handleBack = () => {
+        if (window.history.length > 1) {
+            navigate(-1);
+        } else {
+            // scenario where there's no history
+            navigate('/home');
+        }
+    };
```
Comment (hreis) [inline src/components/Header/index.tsx:20]: added a handleBack function to navigate users to the previous page or home if no history is present, enhancing the navigation experience.

File: src/components/List/index.tsx [hunk 10]
```diff
@@ -7,26 +7,38 @@ import {Container} from './styles';
 interface Props {
     items?: ListItem[];
     hasNavigation?: boolean;
-    isLoading: string;
+    isLoading: boolean;
```
Comment (hreis) [inline src/components/List/index.tsx:10]: isLoading should be boolean

File: src/components/List/index.tsx [hunk 11]
```diff
@@ -7,26 +7,38 @@ import {Container} from './styles';
...
+        );
+    }
+
+    if (!items?.length) {
+        return (
+            <Container>
+                <p>No items to display.</p>
+            </Container>
+        );
+    }
```
Comment (hreis) [inline src/components/List/index.tsx:28]: Added a conditional rendering check to display a message when there are no items, improving user feedback and handling empty states gracefully.

File: src/components/List/index.tsx [hunk 12]
```diff
@@ -7,26 +7,38 @@ import {Container} from './styles';
...
+            {items.map(({url, id, columns, navigationProps}) => (
+                <Card
+                    key={id}
+                    id={id}
+                    columns={columns}
+                    navigationProps={navigationProps}
+                    hasNavigation={hasNavigation}
+                    url={url}
+                />
+            ))}
```
Comment (hreis) [inline src/components/List/index.tsx:41]: Refactored to remove redundant isLoading check and index in the key prop for cleaner, more efficient rendering of list items.

File: src/components/TeamLeadCard/index.tsx [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+        },
+        {
+            key: 'Name',
+            value: `${teamLead.firstName} ${teamLead.lastName}`,
+        },
+        {
+            key: 'Display Name',
+            value: teamLead.displayName,
+        },
+        {
```
Comment (hreis) [inline src/components/TeamLeadCard/index.tsx:23]: team lead card became a component

Comment (hreis) [reply src/components/TeamLeadCard/index.tsx:23]: with his own test file

Comment (hreis) [reply src/components/TeamLeadCard/index.tsx:23]: things are modular :)

File: src/pages/TeamOverview.tsx [hunk 14]
```diff
@@ -1,94 +1,99 @@
...
 import {Container} from '../components/GlobalComponents';
 import Header from '../components/Header';
 import List from '../components/List';
 
-var mapArray = (users: UserData[]) => {
-    return users.map(u => {
-        var columns = [
+const mapItems = (users: UserData[]): ListItem[] => {
+    return users.map(user => {
+        const columns = [
```
Comment (hreis) [inline src/pages/TeamOverview.tsx:14]: be more clear about what variable we have and working with

File: src/pages/TeamOverview.tsx [hunk 15]
```diff
@@ -1,94 +1,99 @@
...
-const TeamOverview = () => {
+const TeamOverview: React.FC = () => {
     const location = useLocation();
-    const {teamId} = useParams();
-    const [pageData, setPageData] = React.useState<PageState>({});
-    const [isLoading, setIsLoading] = React.useState<boolean>(true);
+    const {teamId} = useParams<{ teamId: string }>(); // specify expected params
+    const [pageData, setPageData] = React.useState<TeamOverviewState>({});
+    const [isLoading, setIsLoading] = React.useState(true);
+    const [errorMessage, setErrorMessage] = React.useState('');
```
Comment (hreis) [inline src/pages/TeamOverview.tsx:52]: removed console log so we can have the message error as a state and use this if we want with a toast or something

File: src/pages/TeamOverview.tsx [hunk 16]
```diff
@@ -1,94 +1,99 @@
...
+                    throw new Error('teamId is undefined');
+                }
+                const {teamLeadId, teamMemberIds = []} = await getTeamOverview(teamId);
+                const teamLead = await getUserData(teamLeadId);
+                const teamMembersPromises = teamMemberIds.map(id => getUserData(id));
+                const teamMembers = await Promise.all(teamMembersPromises);
+                setPageData({teamLead, teamMembers});
+            } catch (error) {
+                setErrorMessage(`'Failed to fetch team data: ${error.message}`);
             }
```
Comment (hreis) [inline src/pages/TeamOverview.tsx:68]: Improved async data fetching with better error handling and parallel requests for team member data, enhancing performance and reliability.

File: src/pages/TeamOverview.tsx [hunk 17]
```diff
@@ -1,94 +1,99 @@
...
         <Container>
-            <Header title={`Team ${location.state.name}`} />
-            {!isLoading && mapTLead(pageData.teamLead)}
-            <List items={mapArray(pageData?.teamMembers ?? [])} isLoading={isLoading} />
+            <Header title={`Team ${teamName}`} />
+            <FilterInput
+                value={filter}
+                onChange={handleFilterChange}
+                placeholder="Filter team members..."
+            />
```
Comment (hreis) [inline src/pages/TeamOverview.tsx:93]: new filter component

File: src/pages/TeamOverview.tsx [hunk 18]
```diff
@@ -1,94 +1,99 @@
...
+                value: user.displayName,
             },
             {
                 key: 'Location',
-                value: u.location,
+                value: user.location,
+            },
+            {
+                key: 'Avatar',
+                value: user.avatarUrl,
```
Comment (hreis) [inline src/pages/TeamOverview.tsx:29]: added avatar so i can retrieve the images on each profile card
